docs/blueprints/referral-manager-v1.md

```diff
@@ -0,0 +1,236 @@
+
+## Design
+
+### Definitions
+
+Referral Manager
+: A stand-alone process running outside of satellites. It should only talk to tardigrade branded satellites.
+
+Referral Manager CLI
+: A CLI that will allow operators to invoke referral link generation.
+
+Referral Link
+: A one-time-use URL that contains a unique invitation token.
+
+Invitation Token
+: A random number generated by Referral Manager for each user in the referral program.
+
+Eligible Users
+: Users who have 0 unredeemed referral link.
+
+Owner's Satellite
+: The satellite that an owner of a particular token belongs to.
+
+Redeemed Satellite
+: The satellite where the new user redeeming a token has registered on.
+
+### How it will work
+
+_User Interface_
+
+1. Users will have a referral tab on the UI. On click, it will trigger satellite to send a request to retrieve referral links from Referral Manager:
+    - Referral Manager will query unredeemed tokens in `tokens` table for this user ID and satellite.
+        - if there are unredeemed tokens, Referral Manager respond satellite's request with the tokens.
+        - if no unredeemed token is found for this user ID and such user ID doesn't exist in `users` table, Referral Manager will add an entry for this user ID in `users` table.
+    - Satellite receives the response from Referral Manager.
+        - if the response playload contains tokens, satellite will display them in the UI.
+        - if its an empty response payload, satellite will display a message `No available referral link. Try again later.`
+
+
+_Referral Manager CLI_
+
+1. `referral manager start` initiates the invitation token generation process. This process begins by checking whether there are existing `eligibleUsers` in memory in Referral Manager.
+    - If `eligibleUsers` is empty, the Referral Manager queries `users` table to get users with 0 remaining invites.
+    - If `eligibleUsers` is not empty, it means one of two things:
+        * Someone else is using the Referral Manager at the same time and has not generated/sent tokens yet.
+        * The CLI was terminated between the get `eligibleUsers` step and the generate/send tokens step.
+    
+        Either one of the cases, We will return an error and display `Referral token generation already in progress or canceled early. Do you want to continue anyway?`. If they decide to continue, we call `ClearUsers` to sets eligibleUsers to empty and the Referral Manager queries `users` table to get users with 0 remaining invites. 
+2. The Referral Manager returns the number of users to the CLI.
+3. The CLI receives the number of users and displays: `How many invitation do you want to generate for each user?`
+4. After the user enters the number of invites to generate, the CLI displays: `Generating X amount of invitation tokens for Y amount of users. Yes/No.`
+   - If the input is `Yes`, it will generate x amount of invitation tokens per user based on the input from CLI and save them into `tokens` table.
```

Comment:
Is there a reason we aren't generating them on demand rather than upfront?

e.g. flow wise:
1. referral manager says that you have 3 tokens available,
2. user clicks "create link" or "send link to email",
3. satellite contacts referral manager which creates the token.

Reply:
I think it's because we want to keep a set amount of available tokens we give out to users. If we generate tokens on demand, then we will need to have some sort of global counter to keep track of how many we have generated. Also satellite doesn't know anything about the tokens, so it would have to contact referral manager first to get the token count for a user anyway. If we generate it upfront, then there will be less round trip

Reply:
Not really, each user could have a "tokens available", then the issuing the tokens would be changing just that number.

Reply:
> If we generate tokens on demand, then we will need to have some sort of global counter to keep track of how many we have generated.

Does this means we track the unused token indefinitely?

Reply:
Yea, the marketing team would like to be able to collect metrics on token usage.

Reply:
Updated for generating tokens on demand.